lib/core/src/Cardano/Wallet/Primitive/Scripts.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of wallet's verification key discovery
+-- in a script.
+
+module Cardano.Wallet.Primitive.Scripts
+    ( isShared
+    , retrieveAllVerKeyHashes
+    , toKeyHash
+    ) where
+
+import Prelude
+
+import Cardano.Address.Derivation
+    ( xpubPublicKey )
+import Cardano.Address.Script
+    ( KeyHash (..), Script (..), ScriptHash (..), toScriptHash )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationType (..)
+    , Index (..)
+    , Role (..)
+    , SoftDerivation (..)
+    , hex
+    )
+import Cardano.Wallet.Primitive.AddressDerivation.Shelley
+    ( ShelleyKey (..) )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( SeqState (..), accountPubKey )
+import Crypto.Hash.Utils
+    ( blake2b224 )
+
+import qualified Data.Map.Strict as Map
+import qualified Data.Text.Encoding as T
+
+instance Ord ScriptHash where
+    compare (ScriptHash sh1) (ScriptHash sh2) =
+        compare (T.decodeUtf8 $ hex sh1) (T.decodeUtf8 $ hex sh2)
+
+isShared
+    :: (k ~ ShelleyKey, SoftDerivation k)
+    => Script
+    -> SeqState n k
+    -> ([k 'ScriptK XPub], SeqState n k)
+isShared script s@(SeqState !s1 !s2 !ixs !rpk !prefix !scripts) =
+    let verKeysInScript = retrieveAllVerKeyHashes script
+        toVerKey = deriveAddressPublicKey (accountPubKey s2) MultisigScript
+        minIndex = getIndex @'Soft minBound
+        scriptAddressGap = 10
+        ourVerKeys =
+            map (\ix -> toVerKey (toEnum (fromInteger $ toInteger $ minIndex + ix)))
+            [0 .. scriptAddressGap - 1]
+        ourVerKeyHashesInScript =
+            filter (\keyH -> toKeyHash keyH `elem` verKeysInScript)
+            ourVerKeys
+        toScriptXPub (ShelleyKey k) = ShelleyKey k
+        scriptXPubs = map toScriptXPub ourVerKeyHashesInScript
+    in if null ourVerKeyHashesInScript then
+        ([], s)
```

Comment:
This if is at the wrong level IMO. What about making the intent clearer to the reader:

```hs
let 
    insertIf predicate k v = if predicate v then Map.insert k v else id
in
    ( scriptXPubs
    , SeqState s1 s2 ixs rpk prefix (insertIf (non . null) (toScriptHash script) scriptXPubs scripts)
    )
```

Reply:
done in https://github.com/input-output-hk/cardano-wallet/pull/2308/commits/c92970a14042ece9c693122ccc6064192a8f7d48